### PR TITLE
[CodeGen][MISched] Add misched post-regalloc bottom-up scheduling

### PR DESCRIPTION
**Pull request by michaelmaitland**
This PR contains 2 commits. I plan to commit them separately. 

[CodeGen][MISched] Add misched post-regalloc bottom-up scheduling   

    The main justification for this patch is that I'd like to add a
    HazardRecognizer that does not work in the top-down direction. This creates a
    need for bottom-up scheduling in the PostGenericScheduler. I found that the
    PostGenericScheduler would undo the scheduling which was influenced by the
    HazardRecognition that happend in the pre-regalloc GenericScheduler.

    There is also the possibility that the bottom-up direction will lead to
    performance improvements on certain targets, as this is certainly the case for
    the pre-regalloc GenericScheduler. This patch will give people the
    opportunity to experiment for their sub-targets. However, this patch
    keeps the top-down approach as the default for the PostGenericScheduler

[CodeGen][MISched] dumpSched direction depends on field in DAG.

    This is a precommit to supporting post reg-alloc bottom up scheduling.
    We'd like to have post-ra scheduling direction that can be different from
    pre-ra direction. The current dumpSchedule function is changed in this
    patch to support the fact that the post-ra and pre-ra directions will
    depend on different command line options.

### CODE REVIEW

**Comment by michaelmaitland** [comment]
ping

File: llvm/lib/Target/RISCV/RISCVSchedSiFive7.td [hunk 1]
```diff
@@ -198,6 +198,7 @@ def SiFive7Model : SchedMachineModel {
   let LoadLatency = 3;
   let MispredictPenalty = 3;
   let CompleteModel = 0;
+  let PostRAScheduler = true;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVSchedSiFive7.td:201]: This is conceptually a separate change

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVSchedSiFive7.td]: Updated.

**Comment by michaelmaitland** [comment]
Kindly pinging.

**Comment by michaelmaitland** [comment]
Kindly pinging.

**Comment by wangpc-pp** [comment]
Based on my rough understanding, these changes are really straightforward, so I think this PR makes sense to me.
One thought: the test doesn't show the difference between two directions, is it possible to add a test that will have different scheduled results?

**Comment by michaelmaitland** [comment]
> Based on my rough understanding, these changes are really straightforward, so I think this PR makes sense to me. One thought: the test doesn't show the difference between two directions, is it possible to add a test that will have different scheduled results?

Updated

**Review by wangpc-pp — APPROVED** [review]
I'd like to give a LGTM, but please wait for some days before landing to see if others have comments.

File: llvm/lib/CodeGen/MachineScheduler.cpp [hunk 2]
```diff
@@ -3922,29 +3972,54 @@ void PostGenericScheduler::pickNodeFromQueue(SchedCandidate &Cand) {
...
-      tracePick(Only1, true);
+    if (RegionPolicy.OnlyBottomUp) {
+      assert(!RegionPolicy.OnlyTopDown);
+      SU = Bot.pickOnlyChoice();
+      if (SU) {
+        tracePick(Only1, true);
+      } else {
+        CandPolicy NoPolicy;
+        SchedCandidate BotCand(NoPolicy);
+        // Set the top-down policy based on the state of the current top zone
```
Comment (topperc) [inline llvm/lib/CodeGen/MachineScheduler.cpp:3989]: This is the bottom-up policy isn't it?

Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineScheduler.cpp]: Updated.

**Comment by michaelmaitland** [comment]
Landed in f7cf1f6236ee299d65c2b33429c1d3b729f54c32 and 9106b58ce4e8dada167eec50178a9e154342e4ba

**Pull request closed without merging**